core/src/env.rs

```diff
@@ -185,21 +184,65 @@ impl<E: Ext> LaterExt<E> {
         res
     }
 
-    /// Unset inner ext
-    pub fn take(self) -> Option<E> {
-        self.inner.borrow_mut().take()
+    /// Creates clone for the current reference.
+    ///
+    /// Clone type differs from the [`ExtCarrier`]. For rationale see [`ExtCarrierClone`] docs.
+    pub fn create_carrier_clone(&self) -> ExtCarrierClone<E> {
+        let clone = Self {
+            inner: self.inner.clone(),
+        };
+        ExtCarrierClone(clone)
+    }
+}
+
+/// [`ExtCarrier`]'s clone.
+///
+/// Could be instantiated only by calling [`ExtCarrier::clone`] method.
+///
+/// Carriers of the [`crate::env`] module are actually wrappers over [`Rc`]. If we use [`Rc::clone`] we won't have a guarantee
+/// that [`ExtCarrier::into_inner`] can't be called twice and more on the same data, which potentially leads to panic.
+/// In order to give that guarantee, we mustn't provide an opportunity to unset `Ext` (by calling `into_inner`) on clones.
+/// So this idea is implemented with [`ExtCarrierClone`], which is the clone of [`ExtCarrier`], but with no ability to consume value
+/// to get ownership over the wrapped [`Ext`].
+pub struct ExtCarrierClone<E: Ext>(ExtCarrier<E>);
+
+impl<E: Ext> ExtCarrierClone<E> {
+    /// Calls infallible fn with inner ext
+    pub fn with<R>(&self, f: impl FnOnce(&mut E) -> R) -> Result<R, &'static str> {
+        self.0.with(f)
+    }
+
+    /// Calls fallible fn with inner ext
+    pub fn with_fallible<R>(
+        &self,
+        f: impl FnOnce(&mut E) -> Result<R, &'static str>,
+    ) -> Result<R, &'static str> {
+        self.0.with_fallible(f)
+    }
+}
+
+impl<E: Ext> Clone for ExtCarrierClone<E> {
+    fn clone(&self) -> Self {
```

Comment:
The methos itself is very similar to `create_carrier_clone()`, isn't it?

```suggestion
    fn clone(&self) -> Self { self.0.create_carrier_clone() }
```

Reply:
done in [0817d34](https://github.com/gear-tech/gear/pull/853/commits/0817d34d8fe8168d3d646a8c27dd205e507d585b)